packages/klighd-core/src/depth-map.ts

```diff
@@ -0,0 +1,483 @@
+        let toBeProcessed: Set<Region> = new Set(this.criticalRegions)
+
+        // The regions that have become critical and therfore need to be checked as well
+        let nextToBeProcessed: Set<Region> = new Set()
+
+        while (toBeProcessed.size !== 0) {
+            toBeProcessed.forEach(region => {
+                // Collapse either if the parent region is collapsed or the expansion state changes.
+                if (region.parent && region.parent.expansionState !== Visibility.Expanded) {
+                    this.recursiveCollapseRegion(region, region.parent.expansionState)
+                } else {
+                    const vis = this.getExpansionState(region, viewport, threshold);
+                    if (vis !== Visibility.Expanded) {
+                        if (region.parent) {
+                            nextToBeProcessed.add(region.parent)
+                            this.criticalRegions.add(region.parent)
+                        }
+                        this.recursiveCollapseRegion(region, vis)
+                    } else {
+                        // Add children to check for expansion state change.
+                        region.children.forEach(childRegion => {
+                            childSet.add(childRegion)
+                        })
+                    }
+                }
+            })
+
+            toBeProcessed = nextToBeProcessed;
+            nextToBeProcessed = new Set();
+        }
+
+        // Check all collected child regions of expanded states.
+        childSet.forEach(childRegion => {
+            const vis = this.getExpansionState(childRegion, viewport, threshold);
+            if (vis === Visibility.Expanded) {
+                this.searchUntilCollapse(childRegion, viewport, threshold)
+            } else {
+                if (childRegion.parent) {
+                    // our parent is expanded and we are not, 
+                    // this meand our parent is a critical region
+                    this.criticalRegions.add(childRegion.parent)
+                }
+                this.recursiveCollapseRegion(childRegion, vis)
+            }
+        })
+
+        childSet.clear()
+    }
+
+    /**
+     * Decides the appropriate collapsed or expanded state for a region
+     * based on their size in the viewport and visibility
+     * 
+     * @param region The region in question
+     * @param viewport The currenr viewport
+     * @param threshold The expand/collapse threshold
+     * @returns The appropriate expansion state.
+     */
+    getExpansionState(region: Region, viewport: Viewport, threshold: number): Visibility {
+        // Collapse all invisible states and regions.
+        if (!this.isVisible(region, viewport)) {
+            return Visibility.OutOfBounds
+        } else if (!region.parent) {
+            // Regions without parents should always be expanded if they are visible
+            return Visibility.Expanded
+        } {
```

Comment:
Is there an else missing? Or is this the intended behavior? Either way, just opening a scope like this feels like a bug.

Reply:
Yes an else is missing though as all previous branches diverge by returning this has the same behaviour